File: sdks/typescript/src/snapshot.ts
import { Image } from "./image.js";
import { parseSSEStream } from "./sse.js";

function resolveApiUrl(url: string): string {
  const base = url.replace(/\/+$/, "");
  return base.endsWith("/api") ? base : `${base}/api`;
}

export interface SnapshotInfo {
  id: string;
  orgId: string;
  name: string;
  contentHash: string;
  checkpointId: string;
  status: string;
  manifest: Record<string, unknown>;
  createdAt: string;
  lastUsedAt: string;
}

export interface CreateSnapshotOpts {
  name: string;
  image: Image;
  onBuildLogs?: (log: string) => void;
}

export interface SnapshotOpts {
  apiKey?: string;
  apiUrl?: string;
}

/**
 * Manage pre-built snapshots (named, persistent image checkpoints).
 */
export class Snapshots {
  private apiUrl: string;
  private apiKey: string;

  constructor(opts: SnapshotOpts = {}) {
    this.apiUrl = resolveApiUrl(
      opts.apiUrl ?? process.env.OPENCOMPUTER_API_URL ?? "https://app.opencomputer.dev"
    );
    this.apiKey = opts.apiKey ?? process.env.OPENCOMPUTER_API_KEY ?? "";
  }

  private get headers(): Record<string, string> {
    const h: Record<string, string> = { "Content-Type": "application/json" };
    if (this.apiKey) h["X-API-Key"] = this.apiKey;
    return h;
  }

  /**
   * Create a pre-built snapshot from a declarative image.
   * The server boots a sandbox, runs the image steps, checkpoints it, and stores it under the given name.
   */
  async create(opts: CreateSnapshotOpts): Promise<SnapshotInfo> {
    const headers: Record<string, string> = { ...this.headers };
    if (opts.onBuildLogs) {
      headers["Accept"] = "text/event-stream";
    }

    const resp = await fetch(`${this.apiUrl}/snapshots`, {
      method: "POST",
      headers,
      body: JSON.stringify({
        name: opts.name,
        image: opts.image.toJSON(),
      }),
    });

    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`Failed to create snapshot: ${resp.status} ${text}`);
    }

    if (opts.onBuildLogs && resp.headers.get("content-type")?.includes("text/event-stream")) {
      return parseSSEStream<SnapshotInfo>(resp, opts.onBuildLogs);
    }

    return resp.json();
  }

  /**
   * List all named snapshots for the current org.
   */
  async list(): Promise<SnapshotInfo[]> {
    const resp = await fetch(`${this.apiUrl}/snapshots`, {
      headers: this.headers,
    });

    if (!resp.ok) {
      throw new Error(`Failed to list snapshots: ${resp.status}`);
    }

    return resp.json();
  }

  /**
   * Get a snapshot by name.
   */
  async get(name: string): Promise<SnapshotInfo> {
    const resp = await fetch(`${this.apiUrl}/snapshots/${encodeURIComponent(name)}`, {
      headers: this.headers,
    });

    if (!resp.ok) {
      throw new Error(`Failed to get snapshot: ${resp.status}`);
    }

    return resp.json();
  }

  /**
   * Delete a named snapshot.
   */
  async delete(name: string): Promise<void> {
    const resp = await fetch(`${this.apiUrl}/snapshots/${encodeURIComponent(name)}`, {
      method: "DELETE",
      headers: this.headers,
    });

    if (!resp.ok && resp.status !== 404) {
      throw new Error(`Failed to delete snapshot: ${resp.status}`);
    }
  }
}

File: sdks/python/opencomputer/snapshot.py
"""Snapshot management for OpenSandbox - pre-built named image checkpoints."""

from __future__ import annotations

import os
from typing import Any, Callable

import httpx

from opencomputer.image import Image
from opencomputer.sse import parse_sse_stream


class Snapshots:
    """Manage pre-built snapshots (named, persistent image checkpoints).

    Example::

        snapshots = Snapshots()

        image = (
            Image.base()
            .apt_install(["python3", "python3-pip"])
            .pip_install(["pandas", "numpy"])
        )

        await snapshots.create(name="data-science", image=image)
        info = await snapshots.get("data-science")
        all_snapshots = await snapshots.list()
        await snapshots.delete("data-science")
    """

    def __init__(
        self,
        api_key: str | None = None,
        api_url: str | None = None,
    ) -> None:
        url = api_url or os.environ.get("OPENCOMPUTER_API_URL", "https://app.opencomputer.dev")
        url = url.rstrip("/")
        key = api_key or os.environ.get("OPENCOMPUTER_API_KEY", "")

        api_base = url if url.endswith("/api") else f"{url}/api"

        self._headers: dict[str, str] = {}
        if key:
            self._headers["X-API-Key"] = key

        self._api_base = api_base
        self._client = httpx.AsyncClient(base_url=api_base, headers=self._headers, timeout=300.0)

    async def create(
        self,
        name: str,
        image: Image,
        on_build_logs: Callable[[str], None] | None = None,
    ) -> dict[str, Any]:
        """Create a pre-built snapshot from a declarative image.

        Args:
            name: Unique name for this snapshot.
            image: Declarative Image definition.
            on_build_logs: Optional callback for build log streaming.

        Returns:
            Snapshot info dict with id, name, status, etc.
        """
        body = {"name": name, "image": image.to_dict()}

        # Always use SSE streaming — builds can take minutes and
        # non-streaming requests will hit Cloudflare 524 timeouts.
        headers = {**self._headers, "Accept": "text/event-stream"}
        log_fn = on_build_logs if on_build_logs is not None else lambda _: None
        client = httpx.AsyncClient(
            base_url=self._api_base, headers=headers, timeout=600.0,
        )
        try:
            async with client.stream("POST", "/snapshots", json=body) as resp:
                resp.raise_for_status()
                return await parse_sse_stream(resp, log_fn)
        finally:
            await client.aclose()

    async def list(self) -> list[dict[str, Any]]:
        """List all named snapshots for the current org."""
        resp = await self._client.get("/snapshots")
        resp.raise_for_status()
        return resp.json()

    async def get(self, name: str) -> dict[str, Any]:
        """Get a snapshot by name."""
        resp = await self._client.get(f"/snapshots/{name}")
        resp.raise_for_status()
        return resp.json()

    async def delete(self, name: str) -> None:
        """Delete a named snapshot."""
        resp = await self._client.delete(f"/snapshots/{name}")
        if resp.status_code != 404:
            resp.raise_for_status()

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._client.aclose()
